### PR TITLE
Update UTransport interface specification

### LINKED ISSUE #73: Does UTransport::register_listener match on source or sink address?
The [RegisterListener() spec](https://github.com/eclipse-uprotocol/up-spec/blob/main/up-l1/README.adoc#22-registerlistener) does not explicitly define the (type of) address to match on.

> Register a UListener to receive message(s) for a given UUri (topic).

seems to suggest that the given topic should be matched on a message's destination address (*sink* attribute). However, that would mean that *publish* messages cannot be matched because they only contain an origin address (*source* attribute).

### PR DESCRIPTION
**Pull request by sophokles73**
Changed registerListener method signature to support
matching on message source and sink addresses by means
of UUri patterns.

This change allows UTransport to be used by higher level
APIs for implementing all message exchange patterns
supported by uProtocol.

Fixes #73

### CODE REVIEW

File: up-l1/README.adoc
Comment (stevenhartley) [inline up-l1/README.adoc:219]: Is the new line for each attribute like this a standard for defining APIs?

File: up-l1/README.adoc
Comment (stevenhartley) [inline up-l1/README.adoc:211]: The receive API is a non-blocking synchronous API that returns immediately if there are no messages to consume. I also noticed that I didn't explicitly state that this API is optional to implement only if the transport utilizes pull delivery method.

File: up-l1/README.adoc
Comment (stevenhartley) [inline up-l1/README.adoc:245]: I would like to discuss adding this metadata to the rest of the specs, was this one added to test out the links and then if we agree we would add them to all requirements? How do we link this in our code?

File: up-l2/rpcclient.adoc
Comment (stevenhartley) [inline up-l2/rpcclient.adoc:41]: OUT was meant to be a passed parameter that is an output and not the return value from a method but you prefer to refer to IN as a parameter and out as the return value from the function?

**Review by stevenhartley — CHANGES_REQUESTED** [review]
see comments

File: up-l2/rpcclient.adoc
Comment (sophokles73) [reply up-l2/rpcclient.adoc:41]: To be honest: I have no clue. The semantics of IN, OUT, IN/OUT are, at best, unclear to me.

I had already thought about proposing to use an IDL that has some defined semantics instead of _pseudo code_. Maybe using UML2 semantics or just using Java Interfaces and Methods?

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:245]: Adding these might indeed be a little premature. Ideally, they would serve as OpenFastTrace specification item IDs which could then be included in source code, tests etc to provide for traceability across all artifacts. I haven't yet tested this, I have to admit. In fact, OFT does not (yet) support AsciiDoc files. However, from what I can tell, it should not be too hard to add this to OpenFastTrace.

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:211]: Ok, makes sense. Conversely, does this also mean that the un-/register_listener methods are optional to implement only if the transport utilizes push delivery?

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:219]: I just thought that it would improve readability because you could at once see which are the inputs and the outputs.

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:92]: It is in the Rust library :-)
I wonder if we should actually bother with sync vs async in this spec at all or not leave this up to the language library developers instead.

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:106]: I wonder why I would get back a UStatus instead of Void if sending the message succeeded? FMPOV this would result in a pretty clunky developer experience which would require me to always explicitly check for the UCode to be 0, or am I mistaken?

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:135]: again, in the Rust library, this returns a Future in order to take advantage of async transport client libraries.

File: up-l1/README.adoc
Comment (stevenhartley) [reply up-l1/README.adoc:106]: Clunky yes but simpler than checking if it is void or UStatus etc...

File: up-l1/README.adoc
Comment (stevenhartley) [reply up-l1/README.adoc:92]: Transport I think we need to be explicit as we said historically that send() returning that the message has been delivered to the "next hop". It is only the transport layer acknowledgement and not the communication layer (ex. rpc or pub/sub) acknowledgement.

File: up-l1/README.adoc
Comment (stevenhartley) [reply up-l1/README.adoc:135]: per previous message, the sending of the message is synchronous, not the delivery to the end destination.

File: up-l1/README.adoc
Comment (stevenhartley) [reply up-l1/README.adoc:211]: Yes true

File: up-l1/README.adoc
Comment (stevenhartley) [reply up-l1/README.adoc:245]: I would love that we can do this one a single file (ex. the uri.adoc) and see if this can work end-2-end before we start adding the tags everywhere and then have to remove/change them later.

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:106]: IMHO this
```java
try {
  transport.send(myMessage);
} catch (SendException e) {
  log.error(e);
}
```

is much preferable over this
```java
var result = transport.send(myMessage);
if (result.isError()) {
  log.error("failed to send message: {}", result.getCode());
} else {
  // proceed
}
```
This is also true for Rust and I guess every other language that has an exception mechanism. Using that programming language's exception mechanism seems very desirable to me ...

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:92]: > Transport I think we need to be explicit as we said historically that send() returning that the message has been delivered to the "next hop". It is only the transport layer acknowledgement and not the communication layer (ex. rpc or pub/sub) acknowledgement.

And I think that this is the important thing to make clear. Whether an implementation implements the delivery to the _next hop_ using a blocking/synchronous call or a non-blocking/asynchronous call does not really make a difference regarding the delivery semantics/guarantees, does it?

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:245]: Yes, that is what I have in mind as well. I have spent some time investigating OpenFastTrace's extension API today. Looks very doable ...

File: up-l1/README.adoc
Comment (stevenhartley) [reply up-l1/README.adoc:92]: Whether it uses blocking or not underneath uTransport, the API itself has to be consistent, otherwise we cannot build a generic uStreamer that behaves the same regardless of the underlining transport implementation.

File: up-l1/README.adoc
Comment (PLeVasseur) [reply up-l1/README.adoc:106]: Perhaps a bit of my Rust bias creeping in here, but we have language-level mechanisms to make discriminating success-vs-failure cases in `Result<T, E>`.

### Modeled with explicit success and failure at the type level:
```rust
fn register_listener(
  sourceFilter: UUri,
  sinkFilter: UUri?,
  listener: UListener&
) -> Result<(), UStatus>;
```
allows us to easily:
```rust
let registration_result = client.register_listener(source_filter, sink_filter, listener);

// use pattern matching
match registration_result {
  Err(ustatus) => { /* handle error case with access to `code` and `message` */ }
  Ok(()) => { /* handle success */ }
}

// use if-let pattern
if let Err(ustatus) = registration_result {
  /* handle just the error condition inside of ustatus */
}
```

### Modeled without explicit success and failure at the type level:
```rust
fn register_listener(
  sourceFilter: UUri,
  sinkFilter: UUri?,
  listener: UListener&
) -> UStatus;
```
we then need to:
```rust
let registration_result = client.register_listener(source_filter, sink_filter, listener);

// unpack the UCode and message to then match on and use both
let (code, message) = (registration_result.get_code(), registration_result.get_message());
match code {
  UCode::OK => { /* handle the success case */ }
   _ => { /* handle the remaining cases of `code` and also use `message` */ }
}

// unable to use if-let, must fall back to an unpack and if condition
let (code, message) = (registration_result.get_code(), registration_result.get_message());
if code != UCode::OK {
  /* handle the cases where not a success using `code` and `message` */
}
```

Could we ask for other language implementors to weigh in here with what this would look like?

File: up-l1/README.adoc
Comment (PLeVasseur) [inline up-l1/README.adoc:113]:
```suggestion
For example, if HartleyTransport can map only UPriority to its PDU header, then HartleyTransport would define a PDU payload type that includes UPayload and all the other UAttributes and send that new type in its PDU payload such that the receiver will not lose any metadata. 
```

File: up-l1/README.adoc
Comment (PLeVasseur) [reply up-l1/README.adoc:92]: Agreed, we do need consistency for ability to plug into a uStreamer.

Could we have all the methods of a `UTransport` be asynchronous, giving a bit more control over it to the implementor?

For the record, the `up-streamer-rust` implementation in PR uses the `up-rust` `UTransport` definition which has all async methods.

**Comment by sophokles73** [comment]
@stevenhartley @PLeVasseur I have pushed changes and have amended your comments ...

**Comment by matthewd0123** [comment]
Question/comment. Can/should we also align the error messages returned as well? Good that we're aligning the UCodes, but would ask if it's worth it to align the UStatus messages as well between implementations. Benefit would be consistency, but not sure if it's too much micromanagement.

**Comment by sophokles73** [comment]
> Question/comment. Can/should we also align the error messages returned as well? Good that we're aligning the UCodes, but would ask if it's worth it to align the UStatus messages as well between implementations. Benefit would be consistency, but not sure if it's too much micromanagement.

FMPOV relying on error message format/content is an antipattern that leads to code that is hard to maintain and will produce hard to find bugs during runtime. So, no, I do not think that we should standardize error messages ...

File: up-l1/README.adoc
Comment (PLeVasseur) [inline up-l1/README.adoc:92]: Does this:
```
Otherwise, a failed future which contains a `UCode` indicating the reason for the failure.
```
mean that the [`send` signature](https://github.com/eclipse-uprotocol/up-spec/pull/134/files#diff-a034e4666a20b9cc4b66d777b10e9b4ffac86a7646a7dc79d13c7d067987436fR76) should read:
```
send(message: UMessage) : Future<Void, UCode>
```
?

Comment (sophokles73) [reply up-l1/README.adoc:92]: In Java, for example, the Future<T> class is parameterized with the type of successful completion only because the Future will also be failed with an Exception. Rust is the only language I know of where any type can be used to represent the erroneous outcome.
FMPOV it should be sufficient to define that the Future will be failed with _something_ that contains the relevant information (the UCode in this case). In the Java case this probably means that the Exception contains a UCode typed field. In Rust it might mean that the error type is UCode. Does that make sense?

Comment (PLeVasseur) [reply up-l1/README.adoc:92]: Thanks for the context. Makes sense.

File: up-l1/README.adoc
Comment (PLeVasseur) [inline up-l1/README.adoc:77]: Left over, extra `)`?

File: up-l1/README.adoc
Comment (PLeVasseur) [inline up-l1/README.adoc:140]: Something about the `_` gets mangled or something...?

image

File: up-l1/README.adoc
Comment (PLeVasseur) [inline up-l1/README.adoc:181]: Same thing appears to be happening as [here](https://github.com/eclipse-uprotocol/up-spec/pull/134/commits/03d54831bc214ee82e6fe2ee2895ed387a8ef02a#r1602292080).

File: up-l1/README.adoc
Comment (PLeVasseur) [inline up-l1/README.adoc:216]: Same thing appears to be happening as [here](https://github.com/eclipse-uprotocol/up-spec/pull/134/commits/03d54831bc214ee82e6fe2ee2895ed387a8ef02a#r1602292080).

File: up-l1/README.adoc
Comment (PLeVasseur) [inline up-l1/README.adoc:220]: Can't quite leave a comment that far down, so:

```
* Transport *MUST* support retransmission of messages that are no able to be delivered to the next-hop
```
=>
```
* Transport *MUST* support retransmission of messages that are not able to be delivered to the next-hop
```

File: up-l1/README.adoc
Comment (PLeVasseur) [inline up-l1/README.adoc:191]: Here's a thought -- I think for _most_ folks, they will actually want to block here.

Maybe we can have a short snippet like --

> If you wish to treat this function as synchronous, you may use your language of choice's mechanism for doing so. For example, Java's `CompleteableFuture::complete()` or Rust's `.await`.

Dunno. Too hand-holdy?

@stevenhartley -- sanity check me on the Java point :laughing:

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:140]: That's weird, the vscode preview correctly renders the links and it is the standard AsciiDoc format. Anyway, I will replace it with the (linked to) section title.

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:191]: > Here's a thought -- I think for most folks, they will actually want to block here.

FMPOV the important thing about `receive` is, that it should fail _immediately_ if no message is available, i.e. it should not wait for a message becoming available.

It is not really about whether a particular implementation actually uses async or sync calls. However, IMHO a UTransport implementation should follow a consistent approach across all methods, i.e. either sync or async. If client code needs to be able to deal with async methods anyway, because all of the other methods are async, then I do not see much value in this particular method being the only one defined as sync.

In general, when it comes to IO operations (and UTransport is as close to IO as it gets), using async non-blocking calls has been proven to make systems more performant, scalable and less resource consuming. This also matches my personal experience with other (communication infrastructure) projects ...

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:220]: That whole section needs to be rewritten FMPOV. I'll give it a shot ...

File: up-l1/README.adoc
Comment (sophokles73) [reply up-l1/README.adoc:92]: Maybe we should, instead of using the Future<T> return type, just declare the concrete return type (T) and then also declare an Exception being thrown. Would that help to make it easier to understand? We could then also add some general wording to a section called _Implementation Hints_ or similar, where we advise implementers to use non-blocking async calls. WDYT? @PLeVasseur @stevenhartley

**Comment by sophokles73** [comment]
@stevenhartley @PLeVasseur I have pushed changes as discussed in the call.
Would you mind taking another look?

@tamarafischer I have tried to add some wording to the expected outcome of the send() method. WDYT?

File: up-l1/README.adoc
Comment (PLeVasseur) [reply up-l1/README.adoc]: Please view this as non-blocking for merge. It's a minor point that could be discussed later; more a tip to help a developer, but may just as easily fit in with example code we may write later.

File: up-l1/README.adoc
Comment (PLeVasseur) [reply up-l1/README.adoc]: I like your idea of having a tips & tricks section for implementers in which to put some of this advice. I still think having some level of flexibility _even_ at uP-L1 for languages to write it the most reasonable way makes sense, which is where such a section could bring further value IMHO.

**Comment by PLeVasseur** [comment]
> Would you mind taking another look?

Changes to the section we discussed on the call look good to me!

File: up-l1/README.adoc
Comment (stevenhartley) [inline up-l1/README.adoc:160]:
```suggestion
| The _sink_ address pattern that messages need to match. If omitted, a message **MUST NOT** contain any sink address in order to match.
```

File: up-l2/rpcclient.adoc
Comment (stevenhartley) [inline up-l2/rpcclient.adoc:69]: @sophokles73 can you update this to reflect the UPayload change merged in just now? either we define what UPayload is or we pass payload and UPayloadFormat as 2 separate parameters

File: upclient.adoc
Comment (stevenhartley) [inline upclient.adoc:23]:
```suggestion
A uProtocol Client library is the application facing library that uEntities use to interact with other uEntities. The library contains a programming language and transport protocol specific implementation of the low level link:up-l1/README.adoc[Transport Layer API] and the link:up-l2/rpcclient.adoc[Communication Layer API]. The library also includes implementation specific client-side implementations of the core uServices uDiscovery, uTwin, and uSubscription.
```

File: upclient.adoc
Comment (stevenhartley) [inline upclient.adoc:35]: This will be updated when I add the other items and we will refer to the README I gather (but we need to separate dispatchers from the client library, I'll do that on another pull request)

**Review by stevenhartley — COMMENTED** [review]
Minor changes needed but consider it approved from me once the changes are done. Thanks @sophokles73 !

File: up-l2/rpcclient.adoc
Comment (sophokles73) [reply up-l2/rpcclient.adoc:69]: This will actually need more changes because the CallOptions are gone as well, aren't they? FMPOV this should be done in another PR that deals with the l2 APIs ...

**Comment by sophokles73** [comment]
@stevenhartley I have accepted your proposed changes but in order to merge the PR, GitHub still requires your explicit approval ...

**Review by stevenhartley — APPROVED** [review]
LGTM

**Pull request merged**
